### ISSUE TITLE
RUF012 false positive with annotations set as strings

### ISSUE THREAD

**Issue by ZeeD** [astral-sh/ruff#12288]
minimal case:

```python
from typing import ClassVar
from typing import Final


class C:
    cv_orig: ClassVar[list[int]] = []
    cv_str: 'ClassVar[list[int]]' = []

    cv_final_orig: Final[list[int]] = []
    cv_final_str: 'Final[list[int]]' = []

    use_case: 'Final[list[C]]' = []
```

with `ruff 0.5.1` I got the error
```
RUF012 Mutable class attributes should be annotated with `typing.ClassVar`
```

for `cv_str`, `cv_final_str` and `use_case` (that was my initial scenario...)

**Comment by dhruvmanila** [comment]
I think this is the case for other rules as well like `RUF008`, `RUF009`, etc.

**Comment by charliermarsh** [comment]
👍 Yeah we would need to parse these. I wonder if we should consider parsing these in the parser and adding them to the AST...

**Comment by charliermarsh** [comment]
Actually, we probably can't do that because you often need semantic information to know if a string is a type annotation.

**Comment by dhruvmanila** [comment]
We _do_ parse it: https://github.com/astral-sh/ruff/blob/ecd6865d2800a574baf122583b6e463605af2ef5/crates/ruff_python_parser/src/typing.rs; I guess we need to provide some API to resolve it to the AST node if the model is in a type annotation context.

**Comment by charliermarsh** [comment]
Yeah we just parse it much later, at the end IIRC.

**Comment by charliermarsh** [comment]
Perhaps we could parse these on-demand and store them on the semantic model.

**Comment by charliermarsh** [comment]
I guess that might not work because in theory these need to be evaluated lazily, at the very end (e.g., they can reference symbols that are imported afterwards, IIRC).
